tools/crm_mon.c

```diff
@@ -177,6 +164,267 @@ crm_trigger_t *refresh_trigger = NULL;
+    if (output_format != mon_output_console) {
+        g_set_error(error, G_OPTION_ERROR, CRM_EX_INVALID_PARAM, "-s requires console output format");
+        return FALSE;
+    }
+
+    output_format = mon_output_monitor;
+    options.one_shot = TRUE;
+    return TRUE;
+}
+
+static gboolean
+no_curses_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
+    if (output_format == mon_output_console) {
+        output_format = mon_output_plain;
+    }
+
+    return TRUE;
+}
+
+#define INDENT "                                    "
+
+/* *INDENT-OFF* */
+static GOptionEntry addl_entries[] = {
+    { "interval", 'i', 0, G_OPTION_ARG_CALLBACK, reconnect_cb,
+      "Update frequency in seconds",
+      "SECONDS" },
+
+    { "one-shot", '1', 0, G_OPTION_ARG_NONE, &options.one_shot,
+      "Display the cluster status once on the console and exit",
+      NULL },
+
+    { "disable-ncurses", 'N', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, no_curses_cb,
+      "Disable the use of ncurses",
+      NULL },
+
+    { "daemonize", 'd', 0, G_OPTION_ARG_NONE, &options.daemonize,
+      "Run in the background as a daemon",
+      NULL },
+
+    { "pid-file", 'p', 0, G_OPTION_ARG_FILENAME, &options.pid_file,
+      "(Advanced) Daemon pid file location",
+      "FILE" },
+
+    { "external-agent", 'E', 0, G_OPTION_ARG_FILENAME, &options.external_agent,
+      "A program to run when resource operations take place",
+      "FILE" },
+
+    { "external-recipient", 'e', 0, G_OPTION_ARG_STRING, &options.external_recipient,
+      "A recipient for your program (assuming you want the program to send something to someone).",
+      "RCPT" },
+
+    { "xml-file", 'x', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_CALLBACK, use_cib_file_cb,
+      NULL,
+      NULL },
+
+    { NULL }
+};
+
+static GOptionEntry display_entries[] = {
+    { "group-by-node", 'n', 0, G_OPTION_ARG_NONE, &options.group_by_node,
+      "Group resources by node",
+      NULL },
+
+    { "inactive", 'r', 0, G_OPTION_ARG_NONE, &options.inactive_resources,
+      "Display inactive resources",
+      NULL },
+
+    { "failcounts", 'f', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, show_failcounts_cb,
+      "Display resource fail counts",
+      NULL },
+
+    { "operations", 'o', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, show_operations_cb,
+      "Display resource operation history",
+      NULL },
+
+    { "timing-details", 't', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, show_operations_cb,
+      "Display resource operation history with timing details",
+      NULL },
+
+    { "tickets", 'c', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, show_tickets_cb,
+      "Display cluster tickets",
+      NULL },
+
+    { "watch-fencing", 'W', 0, G_OPTION_ARG_NONE, &options.watch_fencing,
+      "Listen for fencing events. For use with --external-agent",
+      NULL },
+
+    { "fence-history", 'm', 0, G_OPTION_ARG_INT, &options.fence_history_level,
```

Comment:
We'll at least need a callback with G_OPTION_FLAG_OPTIONAL_ARG. Is there a way with GOption to tell whether the argument was specified at all?

If -m wasn't given at all, fence_history_level needs to default to 1 (it's initialized to 2 above). If -m is given without an optional argument, fence_history_level needs to default to 2. And of course it can be explicitly given a value like -m 0.

Reply:
Yes, that can be done with a callback.